### ISSUE TITLE
Roslyn doesn't recognize pure null checks if `default` / `default(Nullable<T>)` or optional parameters are used in Equals calls

### ISSUE THREAD

**Issue by TessenR** [dotnet/roslyn#50844]
**Version Used**: 
```
Branch master (19 Jan 2021)
Latest commit c36b5d3 by msftbot[bot]:
Merge pull request #50583 from CyrusNajmabadi/completoinCommon

Extract common code in completion providers.
```

**Steps to Reproduce**:

Compile the following code:
```cs
class C1 : System.IEquatable<int?>
{
  static void M1(C1 c1)
  {
    if (c1.Equals(null))
    {
      _ = c1.ToString(); // CS8602 dereference of a possibly null reference
    }
  }
  
  static void M2(C1 c1)
  {
    if (c1.Equals(default))
    {
      _ = c1.ToString(); // no warnings
    }
  }

  static void M3(C1 c1)
  {
    if (c1.Equals(default(int?)))
    {
      _ = c1.ToString(); // no warnings
    }
  }
  
  static void M4(C1 c1)
  {
    if (c1.Equals())
    {
      _ = c1.ToString(); // no warnings (same with 'default' and 'default(int)' optional parameter values
    }
  }

  public bool Equals(int? other = null) => false;
}
```

**Expected Behavior**:
Warnings for `c1.ToString()` in every method.
All the methods are exactly the same.

**Actual Behavior**:
`M1` has a warning but other methods don't

**Comment by huoyaoyuan** [comment]
It looks very dangerous to me to specially treat these methods with arbitrary implementation.
These methods can have `[NotNullWhen(true)]` on their parameters to solve this perfectly.

**Comment by TessenR** [comment]
Roslyn already treats `object.Equals`, `object.ReferenceEquals`, `System.IEquatable<T>.Equals` and `System.IEqualityComparer<T>.Equals` as well-known compiler methods.

One of the reasons for this is that existing annotations actually don't cover the equality methods' contracts. E.g. you suggest `[NotNullWhen(true)] ` but `object.ReferenceEquals(nullable, nullable)` can return `true` if both operands are `null`. Moreover Roslyn can update a state of a variable it thinks isn't nullable to nullable based on a comparison with null e.g.
```
void M(string s)
{
  if (s == null)
  {
    s.ToString(); // warning here
  }
}
```
To do the same if you used `if (object.ReferenceEquals(s, null))` you'll again need a new annotation since the current annotations cannot express this contract.

The compiler team decided that these scenarios aren't worth adding new annotations (here are the LDM notes: https://github.com/dotnet/csharplang/blob/master/meetings/2019/LDM-2019-05-15.md#equality-postconditions )

Since the compiler behavior here is to treat these four methods specifically and this issue is about an inconsistent treating of some arguments to them if you want it to stop doing this I suggest creating a separate issue so it can be discussed there.

**Comment by huoyaoyuan** [comment]
Well I'm suggesting on **instance** equals methods, not static ones.

**Comment by TessenR** [comment]
Yes, instance methods can be handled with `[NotNullWhen(true)]` since the instance receiver can't be null in this case and it will rule out the current updating of the receiver variable to nullable which is indeed silly. Would you mind creating an issue to handle instance methods with the annotation? I'd upvote it!
